src/Compiler/Checking/import.fs

```diff
@@ -43,48 +45,101 @@ type AssemblyLoader =
+    val ty1: TType
+    val ty2: TType
+    val canCoerce: CanCoerce
+    val tcGlobals: TcGlobals
+
+    new (ty1, ty2, canCoerce, tcGlobals) =
+        { ty1 = ty1; ty2 = ty2; canCoerce = canCoerce; tcGlobals = tcGlobals }
+
+    interface System.IEquatable<TTypeCacheKey> with
+        member this.Equals other =
+            let ty1 = this.ty1
+            let ty2 = this.ty2
+            let otherTy1 = stripTyEqns this.tcGlobals other.ty1
+            let otherTy2 = stripTyEqns this.tcGlobals other.ty2
+
+            if stampEquals this.tcGlobals ty1 otherTy1 && stampEquals this.tcGlobals ty2 otherTy2 && this.canCoerce = other.canCoerce then
+                true
+            else
+                (
+                    LanguagePrimitives.PhysicalEquality ty1 otherTy1
+                    && LanguagePrimitives.PhysicalEquality ty2 otherTy2
+                    && this.canCoerce = other.canCoerce
+                )
+
+    override this.Equals other =
+        match other with
+        | :? TTypeCacheKey as p -> (this :> System.IEquatable<TTypeCacheKey>).Equals p
+        | _ -> false
+
+    override this.GetHashCode() : int =
+        let g = this.tcGlobals
+
+        let ty1 = stripTyEqns g this.ty1
+        let ty2 = stripTyEqns g this.ty2
+
+        let ty1Hash = combineHash (hashStamp g ty1) (hashTType g ty1)
+        let ty2Hash = combineHash (hashStamp g ty1) (hashTType g ty2)
+
+        let combined = combineHash (combineHash ty1Hash ty2Hash) (hash this.canCoerce)
+
+        combined
+
 //-------------------------------------------------------------------------
 // Import an IL types as F# types.
-//------------------------------------------------------------------------- 
+//-------------------------------------------------------------------------
 
-/// Represents a context used by the import routines that convert AbstractIL types and provided 
-/// types to F# internal compiler data structures. 
+/// Represents a context used by the import routines that convert AbstractIL types and provided
+/// types to F# internal compiler data structures.
 ///
 /// Also caches the conversion of AbstractIL ILTypeRef nodes, based on hashes of these.
 ///
 /// There is normally only one ImportMap for any assembly compilation, though additional instances can be created
-/// using tcImports.GetImportMap() if needed, and it is not harmful if multiple instances are used. The object 
-/// serves as an interface through to the tables stored in the primary TcImports structures defined in CompileOps.fs. 
+/// using tcImports.GetImportMap() if needed, and it is not harmful if multiple instances are used. The object
+/// serves as an interface through to the tables stored in the primary TcImports structures defined in CompileOps.fs.
 [<Sealed>]
 type ImportMap(g: TcGlobals, assemblyLoader: AssemblyLoader) =
     let typeRefToTyconRefCache = ConcurrentDictionary<ILTypeRef, TyconRef>()
 
+    let typeSubsumptionCache = ConcurrentDictionary<TTypeCacheKey, bool>()
```

Comment:
Reminder - default (larger) initial size?

Reply:
Yes, we probably should, it is expected to have at least `all types we use in overloads` size (I think), since we will probably check that type subsumed by `obj` at least once?

Reply:
Put 1024 by default with processor counts